### ISSUE TITLE
New pandas related errors for pyright 1.1.395

### ISSUE THREAD

**Issue by fraghag** [microsoft/pyright#9998]
We started getting new panda related errors with 1.1.395 that we didn't have for 1.1.394. (pandas==2.2.3)

Minimal code:
``` python
import pandas as pd


def test_function(df: pd.DataFrame):
    df["column"].isin([1, 2, 3])


def test_function2():
    test_function(pd.read_csv("some_path.csv"))
```

This gives the following error:
```
/home/me/test_pyright.py
  /home/me/test_pyright.py:5:18 - error: "isin" is not a known attribute of "None" (reportOptionalMemberAccess)
  /home/me/test_pyright.py:9:19 - error: Argument of type "DataFrame | TextFileReader" cannot be assigned to parameter "df" of type "DataFrame" in function "test_function"
    Type "DataFrame | TextFileReader" is not assignable to type "DataFrame"
      "TextFileReader" is not assignable to "DataFrame" (reportArgumentType)
2 errors, 0 warnings, 0 informations
```

To me these errors don't make sense.

**Comment by itsdani** [comment]
To be honest these look like valid type errors to me:
1. `df["column"]` might be `None` if the column doesn't exist, so `.isin()` might be called on `None`
2. `pd.read_csv()` can return two types, and one of those (`TextFileReader`) is not accepted as the parameter of `test_function(df: pd.DataFrame)`

That said these make pandas quite annoying to work with, and I also don't understand why new errors get reported in `1.1.395`.

**However we also got some new errors**, and these seem like actual false positives to me:
```python
pd.DataFrame().rename(columns={"col_4": "new_col_4"}).drop(["col_3"], axis=1)
```
This now reports 
```
"drop" is not a known attribute of "None" [reportOptionalMemberAccess]
```
I think this happens, because pyright doesn't realize that `.rename(...)` has the default value for `inplace: bool=False`, and can't pick the signature with `inplace: Literal[False]` which version only returns `DataFrame` and not `None`. By explicitly specifying `.rename(..., inplace=False)` this error goes away.
<details>
<summary>
signatures of the rename method
</summary>

```python
(method)
def rename(
    mapper: Renamer | None = ...,
    *,
    index: Renamer | None = ...,
    columns: Renamer | None = ...,
    axis: Axis | None = ...,
    copy: bool = ...,
    inplace: Literal[False] = ...,
    level: Level | None = ...,
    errors: IgnoreRaise = ...
) -> DataFrame: ...

def rename(
    mapper: Renamer | None = ...,
    *,
    index: Renamer | None = ...,
    columns: Renamer | None = ...,
    axis: Axis | None = ...,
    copy: bool = ...,
    inplace: bool = ...,
    level: Level | None = ...,
    errors: IgnoreRaise = ...
) -> (DataFrame | None): ...
```

</details>

However for some reason **this workaround still doesn't work in other cases**:

```python
pd.DataFrame().reset_index(inplace=False).groupby(["col_1"])
```
This still reports 
```
"groupby" is not a known attribute of "None" [reportOptionalMemberAccess]
```
despite the fact that the overloaded method with `inplace: Literal[False]` can only return a `DataFrame` and not `None`. For some reason in this case pyright still can't find the correct signature.
<details>
<summary>
signatures of the reset_index method
</summary>

```python
(method)
def reset_index(
    level: Level | Sequence[Level] = ...,
    *,
    col_level: int | _str = ...,
    col_fill: Hashable = ...,
    drop: _bool = ...,
    inplace: Literal[False] = ...,
    allow_duplicates: _bool = ...,
    names: Hashable | list[HashableT@reset_index] = ...
) -> DataFrame: ...

def reset_index(
    level: Level | Sequence[Level] = ...,
    *,
    drop: _bool = ...,
    inplace: _bool | None = ...,
    col_level: int | _str = ...,
    col_fill: Hashable = ...,
    allow_duplicates: _bool = ...,
    names: Hashable | list[HashableT@reset_index] = ...
) -> (DataFrame | None): ...
```
</details>

These are all new in `1.1.395` and didn't get reported in `1.1.394`.

**Comment by itsdani** [comment]
I just realized that I used an older version of pandas, and with the latest version there is no error reported for `pd.DataFrame().reset_index(inplace=False).groupby(["col_1"])`

But without explicitly filling out the parameters to their defaults, the errors still get reported.

And even with providing the default value explicitly, this still reports an error:
```python
data: pd.DataFrame = pd.read_csv("some_path.csv", iterator=False)
```

<details>
<summary>
read_csv signatures
</summary>

```python
(function)
def read_csv(
    filepath_or_buffer: FilePath | ReadCsvBuffer[bytes] | ReadCsvBuffer[str],
    *,
    sep: str | Literal[_NoDefault.no_default] | None = ...,
    delimiter: str | Literal[_NoDefault.no_default] | None = ...,
    header: int | Sequence[int] | Literal['infer'] | None = ...,
    names: Sequence[Hashable] | Literal[_NoDefault.no_default] | None = ...,
    index_col: IndexLabel | Literal[False] | None = ...,
    usecols: UsecolsArgType[Unknown] = ...,
    dtype: DtypeArg | None = ...,
    engine: CSVEngine | None = ...,
    converters: Mapping[Hashable, (...) -> Unknown] | None = ...,
    true_values: list[Unknown] | None = ...,
    false_values: list[Unknown] | None = ...,
    skipinitialspace: bool = ...,
    skiprows: list[int] | int | ((Hashable) -> bool) | None = ...,
    skipfooter: int = ...,
    nrows: int | None = ...,
    na_values: Hashable | Iterable[Hashable] | Mapping[Hashable, Iterable[Hashable]] | None = ...,
    keep_default_na: bool = ...,
    na_filter: bool = ...,
    verbose: bool | Literal[_NoDefault.no_default] = ...,
    skip_blank_lines: bool = ...,
    parse_dates: bool | Sequence[Hashable] | None = ...,
    infer_datetime_format: bool | Literal[_NoDefault.no_default] = ...,
    keep_date_col: bool | Literal[_NoDefault.no_default] = ...,
    date_parser: ((...) -> Unknown) | Literal[_NoDefault.no_default] = ...,
    date_format: str | dict[Hashable, str] | None = ...,
    dayfirst: bool = ...,
    cache_dates: bool = ...,
    iterator: Literal[False] = ...,
    chunksize: None = ...,
    compression: CompressionOptions = ...,
    thousands: str | None = ...,
    decimal: str = ...,
    lineterminator: str | None = ...,
    quotechar: str = ...,
    quoting: int = ...,
    doublequote: bool = ...,
    escapechar: str | None = ...,
    comment: str | None = ...,
    encoding: str | None = ...,
    encoding_errors: str | None = ...,
    dialect: str | Dialect | None = ...,
    on_bad_lines: ... = ...,
    delim_whitespace: bool | Literal[_NoDefault.no_default] = ...,
    low_memory: bool = ...,
    memory_map: bool = ...,
    float_precision: Literal['high', 'legacy'] | None = ...,
    storage_options: StorageOptions = ...,
    dtype_backend: DtypeBackend | Literal[_NoDefault.no_default] = ...
) -> DataFrame: ...

def read_csv(
    filepath_or_buffer: FilePath | ReadCsvBuffer[bytes] | ReadCsvBuffer[str],
    *,
    sep: str | Literal[_NoDefault.no_default] | None = ...,
    delimiter: str | Literal[_NoDefault.no_default] | None = ...,
    header: int | Sequence[int] | Literal['infer'] | None = ...,
    names: Sequence[Hashable] | Literal[_NoDefault.no_default] | None = ...,
    index_col: IndexLabel | Literal[False] | None = ...,
    usecols: UsecolsArgType[Unknown] = ...,
    dtype: DtypeArg | None = ...,
    engine: CSVEngine | None = ...,
    converters: Mapping[Hashable, (...) -> Unknown] | None = ...,
    true_values: list[Unknown] | None = ...,
    false_values: list[Unknown] | None = ...,
    skipinitialspace: bool = ...,
    skiprows: list[int] | int | ((Hashable) -> bool) | None = ...,
    skipfooter: int = ...,
    nrows: int | None = ...,
    na_values: Hashable | Iterable[Hashable] | Mapping[Hashable, Iterable[Hashable]] | None = ...,
    keep_default_na: bool = ...,
    na_filter: bool = ...,
    verbose: bool | Literal[_NoDefault.no_default] = ...,
    skip_blank_lines: bool = ...,
    parse_dates: bool | Sequence[Hashable] | None = ...,
    infer_datetime_format: bool | Literal[_NoDefault.no_default] = ...,
    keep_date_col: bool | Literal[_NoDefault.no_default] = ...,
    date_parser: ((...) -> Unknown) | Literal[_NoDefault.no_default] = ...,
    date_format: str | dict[Hashable, str] | None = ...,
    dayfirst: bool = ...,
    cache_dates: bool = ...,
    iterator: bool = ...,
    chunksize: int | None = ...,
    compression: CompressionOptions = ...,
    thousands: str | None = ...,
    decimal: str = ...,
    lineterminator: str | None = ...,
    quotechar: str = ...,
    quoting: int = ...,
    doublequote: bool = ...,
    escapechar: str | None = ...,
    comment: str | None = ...,
    encoding: str | None = ...,
    encoding_errors: str | None = ...,
    dialect: str | Dialect | None = ...,
    on_bad_lines: ... = ...,
    delim_whitespace: bool | Literal[_NoDefault.no_default] = ...,
    low_memory: bool = ...,
    memory_map: bool = ...,
    float_precision: Literal['high', 'legacy'] | None = ...,
    storage_options: StorageOptions = ...,
    dtype_backend: DtypeBackend | Literal[_NoDefault.no_default] = ...
) -> (DataFrame | TextFileReader): ...
```


</details>

**Comment by itsdani** [comment]
I think this might be a duplicate of #9984, or at the very least also caused by the fix for #9901

**Comment by erictraut** [comment]
Yes, this is a duplicate of https://github.com/microsoft/pyright/issues/9984.
